File: kinect-input/src/sources/mock_source.cpp
#include "sources/mock_source.h"

#include <chrono>
#include <cmath>
#include <cstdio>
#include <numbers>
#include <thread>
#include <utility>

MockSource::MockSource(bool realtime)
    : realtime_(realtime), rng_(std::random_device{}()) {
    say(">> Simulation: Person steht still (Kalibrierung)...");
    hold(1.5);
    say(">> Simulation: SPRUNG");
    jump();
    hold(1.0);
    say(">> Simulation: Schritt nach LINKS");
    step_to(-kSideX);
    hold(0.8);
    say(">> Simulation: zurück zur Mitte");
    step_to(0.0);
    hold(0.8);
    say(">> Simulation: DUCKEN (1.2s)");
    hold(1.2, kCrouchHeight);
    say(">> Simulation: wieder aufstehen");
    hold(1.0);
    say(">> Simulation: Schritt nach RECHTS");
    step_to(+kSideX);
    hold(0.8);
    say(">> Simulation: zurück zur Mitte");
    step_to(0.0);
    hold(0.5);
    say(">> Simulation: noch ein SPRUNG");
    jump();
    hold(1.0);
    end_message_ = ">> Simulation beendet.";
}

double MockSource::noise() {
    // Sensor-Rauschen wie im Python-Prototyp: gleichverteilt in [-0.01, +0.01]
    return std::uniform_real_distribution<double>(-0.01, 0.01)(rng_);
}

void MockSource::say(std::string message) {
    pending_message_ = std::move(message);
}

void MockSource::push(double x, double height) {
    script_.push_back(Step{std::move(pending_message_), x + noise(), height + noise()});
    pending_message_.clear();
}

void MockSource::hold(double seconds, double height) {
    const int n = static_cast<int>(seconds * kFps);
    for (int i = 0; i < n; ++i) {
        push(x_, height);
    }
}

void MockSource::jump(double duration) {
    const int n = static_cast<int>(duration * kFps);
    for (int i = 0; i < n; ++i) {
        const double p = static_cast<double>(i) / (n - 1);
        // Parabel rauf und runter
        push(x_, kStandHeight + kJumpPeak * std::sin(std::numbers::pi * p));
    }
}

void MockSource::step_to(double target_x, double duration) {
    const int n = static_cast<int>(duration * kFps);
    const double start = x_;
    for (int i = 0; i < n; ++i) {
        x_ = start + (target_x - start) * (i + 1) / n;
        push(x_, kStandHeight);
    }
}

bool MockSource::next(BodyState& out) {
    if (index_ >= script_.size()) {
        if (!end_message_.empty()) {
            std::printf("%s\n", end_message_.c_str());
            end_message_.clear();
        }
        return false;
    }

    const Step& step = script_[index_++];
    if (!step.message.empty()) {
        std::printf("%s\n", step.message.c_str());
    }
    out = BodyState{step.x, step.height, t_};
    t_ += kDt;

    if (realtime_) {
        std::this_thread::sleep_for(std::chrono::duration<double>(kDt));
    }
    return true;
}

File: kinect-input-rust/src/sources/mock_source.rs
use std::f64::consts::PI;
use std::thread;
use std::time::Duration;

use crate::body_state::BodyState;

use super::Source;

const FPS: f64 = 30.0;
const DT: f64 = 1.0 / FPS;
const STAND_HEIGHT: f64 = 1.75;
const CROUCH_HEIGHT: f64 = 1.30;
const JUMP_PEAK: f64 = 0.32;
const SIDE_X: f64 = 0.55;

#[derive(Debug, Clone)]
struct Step {
    message: Option<String>,
    x: f64,
    height: f64,
}

#[derive(Debug)]
pub struct MockSource {
    script: Vec<Step>,
    index: usize,
    t: f64,
    x: f64,
    pending_message: Option<String>,
    end_message: Option<String>,
    realtime: bool,
    rng: Lcg,
}

impl Default for MockSource {
    fn default() -> Self {
        Self::new(true)
    }
}

impl MockSource {
    pub fn new(realtime: bool) -> Self {
        let mut source = Self {
            script: Vec::new(),
            index: 0,
            t: 0.0,
            x: 0.0,
            pending_message: None,
            end_message: None,
            realtime,
            rng: Lcg::new(0x5eed_1234),
        };

        source.say(">> Simulation: Person steht still (Kalibrierung)...");
        source.hold(1.5, STAND_HEIGHT);
        source.say(">> Simulation: SPRUNG");
        source.jump(0.45);
        source.hold(1.0, STAND_HEIGHT);
        source.say(">> Simulation: Schritt nach LINKS");
        source.step_to(-SIDE_X, 0.35);
        source.hold(0.8, STAND_HEIGHT);
        source.say(">> Simulation: zurueck zur Mitte");
        source.step_to(0.0, 0.35);
        source.hold(0.8, STAND_HEIGHT);
        source.say(">> Simulation: DUCKEN (1.2s)");
        source.hold(1.2, CROUCH_HEIGHT);
        source.say(">> Simulation: wieder aufstehen");
        source.hold(1.0, STAND_HEIGHT);
        source.say(">> Simulation: Schritt nach RECHTS");
        source.step_to(SIDE_X, 0.35);
        source.hold(0.8, STAND_HEIGHT);
        source.say(">> Simulation: zurueck zur Mitte");
        source.step_to(0.0, 0.35);
        source.hold(0.5, STAND_HEIGHT);
        source.say(">> Simulation: noch ein SPRUNG");
        source.jump(0.45);
        source.hold(1.0, STAND_HEIGHT);
        source.end_message = Some(">> Simulation beendet.".to_string());

        source
    }

    fn say(&mut self, message: &str) {
        self.pending_message = Some(message.to_string());
    }

    fn push(&mut self, x: f64, height: f64) {
        let message = self.pending_message.take();
        let nx = x + self.noise();
        let nh = height + self.noise();
        self.script.push(Step {
            message,
            x: nx,
            height: nh,
        });
    }

    fn hold(&mut self, seconds: f64, height: f64) {
        let frames = (seconds * FPS) as usize;
        for _ in 0..frames {
            self.push(self.x, height);
        }
    }

    fn jump(&mut self, duration: f64) {
        let frames = (duration * FPS) as usize;
        for i in 0..frames {
            let p = i as f64 / (frames - 1) as f64;
            self.push(self.x, STAND_HEIGHT + JUMP_PEAK * (PI * p).sin());
        }
    }

    fn step_to(&mut self, target_x: f64, duration: f64) {
        let frames = (duration * FPS) as usize;
        let start = self.x;
        for i in 0..frames {
            self.x = start + (target_x - start) * (i + 1) as f64 / frames as f64;
            self.push(self.x, STAND_HEIGHT);
        }
    }

    fn noise(&mut self) -> f64 {
        self.rng.next_range(-0.01, 0.01)
    }
}

impl Source for MockSource {
    fn next(&mut self) -> Option<BodyState> {
        if self.index >= self.script.len() {
            if let Some(message) = self.end_message.take() {
                println!("{message}");
            }
            return None;
        }

        let step = &self.script[self.index];
        self.index += 1;

        if let Some(message) = &step.message {
            println!("{message}");
        }

        let out = BodyState::new(step.x, step.height, self.t);
        self.t += DT;

        if self.realtime {
            thread::sleep(Duration::from_secs_f64(DT));
        }

        Some(out)
    }
}

#[derive(Debug)]
struct Lcg {
    state: u64,
}

impl Lcg {
    fn new(seed: u64) -> Self {
        Self { state: seed }
    }

    fn next_range(&mut self, min: f64, max: f64) -> f64 {
        self.state = self
            .state
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1);
        let unit = ((self.state >> 11) as f64) / ((1u64 << 53) as f64);
        min + (max - min) * unit
    }
}
